server/src/daemon/prompter.rs

```diff
@@ -0,0 +1,210 @@
+// org.gnome.keyring.Prompter
+// https://gitlab.gnome.org/GNOME/gcr/-/blob/master/gcr/org.gnome.keyring.Prompter.xml
+
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use tokio;
+use zbus::{
+    fdo, interface,
+    message::Header,
+    proxy,
+    zvariant::{
+        self, DeserializeDict, ObjectPath, OwnedObjectPath, OwnedValue, SerializeDict, Type,
+    },
+    Connection, SignalContext,
+};
+
+use super::{secret_exchange, service_manager::ServiceManager};
+
+// May be change this to /org/oo7_daemon/Prompt
+const SECRET_PROMPTER_PREFIX: &str = "/org/gnome/keyring/Prompt/";
+
+#[derive(Clone, Default, DeserializeDict, Debug, Type, SerializeDict)]
+#[zvariant(signature = "dict")]
+pub struct Properties {
+    title: Option<String>,
+    message: Option<String>,
+    description: Option<String>,
+    warning: Option<String>,
+    #[zvariant(rename = "choice-label")]
+    choice_label: Option<String>,
+    #[zvariant(rename = "caller-window")]
+    caller_window: Option<String>,
+    #[zvariant(rename = "continue-label")]
+    continue_label: Option<String>,
+    #[zvariant(rename = "cancel-label")]
+    cancel_label: Option<String>,
+    #[zvariant(rename = "choice-chosen")]
+    choice_chosen: Option<bool>,
+    #[zvariant(rename = "password-new")]
+    password_new: Option<bool>,
+    #[zvariant(rename = "password-strength")]
+    password_strength: Option<u32>,
+}
+
+// org.gnome.keyring.internal.Prompter.Callback
+
+#[derive(Clone, Debug)]
+pub struct PrompterCallback {
+    properties: Arc<Mutex<Properties>>,
+    path: OwnedObjectPath,
+}
+
+#[interface(name = "org.gnome.keyring.internal.Prompter.Callback")]
+impl PrompterCallback {
+    pub async fn prompt_ready(
+        &self,
+        reply: &str,            // the purpose of this?
+        properties: Properties, // this should probably be &mut Properties
+        exchange: &str,         // this should probably be &mut String
+        #[zbus(connection)] connection: &zbus::Connection,
+        #[zbus(header)] header: Header<'_>,
+    ) -> fdo::Result<()> {
+        println!("prompter: {}", header.path().unwrap());
+
+        let mut properties = properties.clone();
+
+        if header.path().unwrap().as_str().to_string().contains("/u") {
```

Comment:
By having different constructors, you can easily find out this. Even create a different Properties type based on that.

I would suggest defining an enum like this
```rust
pub enum PromptSource {
  Unlock,
  NewCollection,
}
```
You can then add a `Properties::for_source` that generates this. Or a `PromptSource::path(counter)`

Reply:
Done, thanks!